fhir-codegen/gradle-plugin/src/main/kotlin/com/google/fhir/codegen/EnumTypeSpecGenerator.kt

```diff
@@ -0,0 +1,288 @@
+                    addStatement(
+                      "else -> throw IllegalArgumentException(\"Unknown code \$code for enum %L\")",
+                      enumClassName,
+                    )
+                  }
+                  .endControlFlow()
+                  .build()
+              )
+              .build()
+          )
+        }
+        .build()
+    return typeSpec
+  }
+
+  /**
+   * Instantiate a [FhirEnum] to facilitate the generation of Kotlin enum classes. The enum
+   * constants are derived from concepts defined in a `ValueSet`, a `CodeSystem`, or both. When both
+   * are provided, the constants are generated by intersecting the concepts from the `ValueSet` and
+   * the `CodeSystem`. If only one is provided, the constants are based solely on that source. Any
+   * nested concepts are flattened, transformed, and included in the resulting enum constants.
+   *
+   * NOTE: This excludes all systems that equals "http://unitsofmeasure.org" or start with "urn",
+   * e.g. urn:ietf:bcp:13, urn:ietf:bcp:47,urn:iso:std:iso:4217 typically used for MIMETypes,
+   * Currency Code etc.
+   */
+  private fun generateEnum(valueSet: ValueSet, codeSystemMap: Map<String, CodeSystem>): FhirEnum? {
+    return valueSet.compose
+      ?.include
+      ?.filter { it.isValueSystemSupported() }
+      ?.let {
+        val enumConstants = mutableListOf<FhirEnumConstant>()
+        for (include in it) {
+          val system = include.system!!
+          val codeSystem = codeSystemMap[system]
+
+          generateEnumConstants(
+            system = system,
+            codeSystemConcepts = codeSystem?.concept,
+            valueSetConcepts = include.concept,
+            enumConstants = enumConstants,
+          )
+        }
+        return FhirEnum(valueSet.description, enumConstants)
+      }
+  }
+
+  private fun generateEnumConstants(
+    system: String,
+    codeSystemConcepts: List<Concept>?,
+    valueSetConcepts: List<Concept>?,
+    enumConstants: MutableList<FhirEnumConstant>,
+  ) {
+    if (!codeSystemConcepts.isNullOrEmpty() && !valueSetConcepts.isNullOrEmpty()) {
+      val expectedCodesSet = valueSetConcepts.mapTo(hashSetOf()) { it.code }
+      generateNestedEnumConstants(valueSetConcepts, system, enumConstants, expectedCodesSet)
+    } else if (!valueSetConcepts.isNullOrEmpty() && codeSystemConcepts.isNullOrEmpty()) {
+      generateNestedEnumConstants(valueSetConcepts, system, enumConstants, null)
+    } else if (!codeSystemConcepts.isNullOrEmpty() && valueSetConcepts.isNullOrEmpty()) {
+      generateNestedEnumConstants(codeSystemConcepts, system, enumConstants, null)
+    }
+  }
+
+  private fun generateNestedEnumConstants(
+    concepts: List<Concept>,
+    system: String,
+    enumConstants: MutableList<FhirEnumConstant>,
+    expectedCodesSet: Set<String>?,
+  ) {
+    val arrayDeque: ArrayDeque<Concept> = ArrayDeque(concepts)
+
+    while (arrayDeque.isNotEmpty()) {
```

Comment:
again, can we make this more "functional" and avoid using while?

Reply:
Oh, I see. I missed the context. This can probably be refactored.